fix: breadcrumb links without `href` should not have hover styles

srmagura opened the pull request:
Breadcrumb items do not necessarily have an `href`, but they always change color when the user hovers over them. The hover styles give the impression that the breadcrumb item is a clickable link even if this is not the case.

This can be seen in the [Redux documentation](https://redux.js.org/tutorials/essentials/part-1-overview-concepts) — neither "Tutorials" nor "Redux Essentials" are links, yet they have hover styles.

This PR uses the `:link` pseudoselector to fix this.

Edit: Signed the CLA.

slorber commented on packages/core/styles/components/breadcrumb.pcss:
```diff
@@ -71,7 +71,7 @@
     @mixin transition background color;
 
-    &:hover {
+    &:link:hover {
```
> The :link CSS pseudo-class represents an element that has not yet been visited.

Does it mean that we'll lose the hover effect once the link is visited? 😅 

Maybe using `:is(a):hover` would be better? (90% support good enough for this usecase)

TIL there's also an `:any-link` pseudo-selector, looks like a potential good fit?

slorber replied:
yes, not very intuitive 😅

slorber commented on packages/core/styles/content/typography.pcss:
```diff
@@ -41,7 +41,7 @@ a {
   @mixin transition color;
 
-  &:hover {
+  &:link:hover {
```
🤷‍♂️ this one looks more general and not sure it's directly related to your problem, considering the element is already a `<a>` there?

Can it have unwanted side effects? + not sure `:link` is what we want anyway

srmagura replied:
You can have `<a>` elements that don't represent links (like headings). It usually only makes sense to display hover styles if the `<a>` represents a link.

slorber reviewed (changes requested):
Agree this UX is not ideal and we should change it.

> Breadcrumb items do not necessarily have an href, but they always change color when the user hovers over them. 

In practice it's more:

> breadcrumb items might be simple `<span>` instead of `<a>` when we don't have a link

Not 100% sure using `:link` is the solution here?

---

Also worth adding a test to our demo page here: https://deploy-preview-266--infima.netlify.app/demo/

We only have breadcrumb items with links currently (unlike what we have on the Docusaurus side in practice). Please try to add something here that helps reviewing the current PR.

srmagura commented:
Thanks for the feedback. I switched to `:any-link` and added an example.

> breadcrumb items might be simple <span> instead of <a> when we don't have a link

True, but I assume you would not want hover styles if the breadcrumb item is a `<span>`, correct?

Josh-Cena commented:
```diff
  diff --git a/packages/core/dist/css/default/default.css b/packages/core/dist-branch/css/default/default.css
  index 95fc6e3..f258e29 100644
  --- a/packages/core/dist/css/default/default.css
  +++ b/packages/core/dist-branch/css/default/default.css
  @@ -1233,[7](https://github.com/facebookincubator/infima/runs/7197269101?check_suite_focus=true#step:4:7) +1233,25 @@ a {
     transition: color var(--ifm-transition-fast) var(--ifm-transition-timing-default);
   }
  
  -a:hover {
  +a:link:hover, a:visited:hover {
  +    color: var(--ifm-link-hover-color);
  +    /* autoprefixer: ignore next */
  +    text-decoration: var(--ifm-link-hover-decoration);
  +  }
  +
  +a:-webkit-any-link:hover {
  +    color: var(--ifm-link-hover-color);
  +    /* autoprefixer: ignore next */
  +    text-decoration: var(--ifm-link-hover-decoration);
  +  }
  +
  +a:-moz-any-link:hover {
  +    color: var(--ifm-link-hover-color);
  +    /* autoprefixer: ignore next */
  +    text-decoration: var(--ifm-link-hover-decoration);
  +  }
  +
  +a:any-link:hover {
       color: var(--ifm-link-hover-color);
       /* autoprefixer: ignore next */
       text-decoration: var(--ifm-link-hover-decoration);
  @@ -172[8](https://github.com/facebookincubator/infima/runs/7197269101?check_suite_focus=true#step:4:8),7 +1746,22 @@ hr {
       transition-timing-function: var(--ifm-transition-timing-default);
     }
  
  -.breadcrumbs__link:hover {
  +.breadcrumbs__link:link:hover, .breadcrumbs__link:visited:hover, area.breadcrumbs__link[href]:hover {
  +      background: var(--ifm-breadcrumb-item-background-active);
  +      text-decoration: none;
  +    }
  +
  +.breadcrumbs__link:-webkit-any-link:hover {
  +      background: var(--ifm-breadcrumb-item-background-active);
  +      text-decoration: none;
  +    }
  +
  +.breadcrumbs__link:-moz-any-link:hover {
  +      background: var(--ifm-breadcrumb-item-background-active);
  +      text-decoration: none;
  +    }
  +
  +.breadcrumbs__link:any-link:hover {
         background: var(--ifm-breadcrumb-item-background-active);
         text-decoration: none;
       }
  ```

srmagura commented:
@Josh-Cena Why are those changes necessary? I tested in Chrome, Safari, and Firefox and the vendor prefixes were not necessary.

Josh-Cena commented:
Ah, no, that's the "post diff comment" workflow's result 😄 It can't be posted on PRs created from forks, so I manually post it so we know how the distributed CSS has changed. See https://github.com/facebookincubator/infima/pull/255#issuecomment-1120486531 for example.

srmagura commented:
Ooooohhhhh, OK. Do I need to do anything to get that workflow to pass?

Josh-Cena commented:
No, it's a problem with our workflow plus some permission quirks. It's mostly okay this way.

slorber commented:
LGTM thanks 👍

pranabdas commented on packages/core/styles/content/typography.pcss:
```diff
@@ -41,7 +41,7 @@ a {
+  &:any-link:hover {
     color: var(--ifm-link-hover-color);
     /* autoprefixer: ignore next */
     text-decoration: var(--ifm-link-hover-decoration);
```
I guess this is causing `text-decoration: underline` for all links including buttons, sidebar menu, navbar, navbar dropdown menu. https://github.com/facebook/docusaurus/issues/7748

yangshun replied:
I think adding this pseudo selector increases the specificity

srmagura replied:
Thanks for pointing this out.

https://github.com/facebookincubator/infima/pull/267
